File: 백준/Silver/12891. DNA 비밀번호/DNA 비밀번호.cs
var input = Array.ConvertAll(Console.ReadLine().Split(), int.Parse);
var str = Console.ReadLine();
var ACGT = Array.ConvertAll(Console.ReadLine().Split(), int.Parse);

var cnt = new int[4];
int l = 0, r = input[1], answer = 0;

var getCount = (string str, int n) =>
{
    foreach(var c in str)
    {
        cnt[c switch { 'A' => 0, 'C' => 1, 'G' => 2, _ => 3 }] += n;
    }
};

var IsValid = (int[] arr) => ACGT.Zip(arr, (a, b) => a > b).Count(x => x == true) == 0;

getCount(str[l..r], 1);
if (IsValid(cnt)) answer++;

while(r < input[0])
{
    getCount(str[l++].ToString(), -1);
    getCount(str[r++].ToString(), 1);
    if (IsValid(cnt)) answer++;
}

Console.WriteLine(answer);

File: 백준/Silver/12891. DNA 비밀번호/DNA 비밀번호.cc
#include <iostream>

using namespace std;

int condi[4];
int cnt[4];

int main()
{
    string str;
    int l, r, length, size, count = 0;

    auto GetCount = [](int *cnt, char det, int n)
    {
        switch (det)
        {
        case 'A':
            cnt[0] += n;
            break;
        case 'C':
            cnt[1] += n;
            break;
        case 'G':
            cnt[2] += n;
            break;
        case 'T':
            cnt[3] += n;
            break;
        }
    };
    auto IsEnd = [](int *arr1, int *arr2, int length)
    {
        for (int i = 0; i < length; i++)
        {
            if (arr1[i] > arr2[i])
                return false;
        }
        return true;
    };

    cin >> size >> length;
    cin >> str;
    cin >> condi[0] >> condi[1] >> condi[2] >> condi[3];

    for (int i = 0; i < length; i++)
        GetCount(cnt, str[i], 1);

    l = 0, r = length - 1;
    while (1)
    {
        if (IsEnd(condi, cnt, 4))
            count++;
        if (++r >= size)
            break;
        GetCount(cnt, str[l++], -1);
        GetCount(cnt, str[r], 1);
    }

    cout << count;

    return 0;
}